Display a warning when there are multiple documents or folders of the same name

jonpo opened the pull request:
@isaacwhite 
### Description of Change
Add handling to detect duplicate slugs in `buildTreeFromData`. Flag duplicate resources and use that flag to display an error message when rendering a response. This doesn't address the problem where if there are duplicates Library doesn't consistently show one of them––wasn't sure the best way to go about that. Let me know if there's a better way of doing this that I didn't see!

Because of the way file fixtures are being generated, adding to them (in order to test the duplicate scenario) changed some fixture identifiers, which is why I had to update so many tests here. It might be worth changing the way those fixtures get generated so that adding to them in the future doesn't change the identifiers?

### Related Issue
https://github.com/nytimes/library/issues/4

### Screenshots
<img width="1333" alt="Screen Shot 2019-07-16 at 5 06 28 PM" src="https://user-images.githubusercontent.com/6548331/61402600-7523eb00-a8a1-11e9-88fd-8e721d779870.png">
<img width="1333" alt="Screen Shot 2019-07-17 at 1 08 44 PM" src="https://user-images.githubusercontent.com/6548331/61402607-78b77200-a8a1-11e9-86dc-3af312ab4010.png">


### Motivation and Context
Let Library users know that the contents of a folder are not accurately represented in Library if there are multiple documents or folders of the same name.

### Checklist
<!-- Cross out items that don't apply and leave a short description of why the item isn't relevant. Check off ([x]) items you complete. -->

- [x] Ran `npm run lint` and updated code style accordingly
- [x] `npm run test` passes
- [x] PR has a description and all contributors/stakeholder are noted/cc'ed
- [x] tests are updated and/or added to cover new code
- [ ] relevant documentation is changed and/or added

isaacwhite commented on layouts/partials/warning.ejs:
```diff
@@ -0,0 +1,3 @@
+<div class="warning">
+  <%= message %>
```
Since this template is named `warning`, would it make sense to include the `Warning:` as part of the template?

jonpo replied:
👍

isaacwhite commented on server/list.js:
```diff
@@ -235,8 +235,14 @@ function buildTreeFromData(rootParent, previousData, breadcrumb) {
+      // recurse building up breadcrumb
+      memo.children[slug] = buildTreeFromData(id, previousData, nextCrumb)
+    } else {
+      log.debug(`Folder ${parentInfo.name} contains duplicate resources with slug ${slug}`)
```
How would you feel about this as a `warn` level log, since we shouldn't really encourage this case to persist?

jonpo replied:
👍

isaacwhite commented on config/strings.yaml:
```diff
@@ -80,3 +80,8 @@ error:
 
+# warning messages
+warning:
+  duplicateDoc: "Warning: Multiple documents in this folder share the same name as this document. Only one of the documents will be accesible through Library."
```
It looks like we don't know for sure whether the other item in the drive that has the exact same slug as this item is the same type; you could mix and match.

A little more complex than a static warning message, but what if we made the warning include the names of the other items that would have shown up at this location to make it easier to resolve the issue? We could save the names of the other documents (or folders!) where we currently set  `isDuplicate`, then pass them into a yaml function like in the `moveFile` definition. In that case, we might want to discard the boolean and instead create an array called `duplicates` as a property.

The general logic of the diff wouldn't change much, but we could drop the ternary in `nav.ejs` and be more specific without needing as many configuration options in strings.yaml.

jonpo replied:
Good idea! I'll make the changes.

jonpo commented:
@isaacwhite I addressed your comments. Now the error message looks like this:
```Warning: Multiple resources in this folder share the same name: duplicate document, duplicate document. Only one will be accesible through Library.```
Do you think it would be more helpful if we displayed the names as links to the actual drive resources?

isaacwhite commented on server/routes/categories.js:
```diff
@@ -44,7 +44,8 @@ async function handleCategory(req, res) {
     id,
-    template: stringTemplate
+    template: stringTemplate,
+    duplicateNames: duplicates ? duplicates.join(', ') : ''
```
What do you think about passing the `duplicates` directly into the render call so that future changes wouldn't touch this file? Then we could perform the `.join()` inside the strings.yaml itself.

jonpo replied:
👍

isaacwhite commented on server/list.js:
```diff
@@ -235,8 +235,17 @@ function buildTreeFromData(rootParent, previousData, breadcrumb) {
+      log.warn(`Folder ${parentInfo.name} contains duplicate resources with slug ${slug}`)
+      const { name } = docsInfo[id]
+      // record duplicate resources so that a warning can be displayed
+      memo.children[slug].duplicates = memo.children[slug].duplicates
```
Could go either way, but would this be more explicit?
```
const previousDupes = memo.children[slug].duplicates || []
memo.children[slug].duplicates = previousDupes.concat(name)
```

jonpo replied:
Yeah I like that 👍

jonpo commented on config/strings.yaml:
```diff
@@ -80,3 +80,7 @@ error:
 
+# warning messages
+warning:
+  duplicate: !!js/function (names) => `Multiple resources in this folder share the same name${':' + ' ' + names}. Only one will be accesible through Library.`
```
That's what I originally had, but I was getting this error:
```
[0]   throw generateError(state, message);
[0]   ^
[0] YAMLException: cannot resolve a node with !<tag:yaml.org,2002:js/function> explicit tag at line 85, column 93:
[0]      ...  this folder share the same name: ${names}. Only one will be acc ...
```
Tried googling around a bit but I didn't find anything useful. What I have here is the only variation I tried that didn't throw the error 🙁

isaacwhite replied:
Ah, I see. It's getting upset about a `:` character as part of a key not enclosed in quotes. Two other options I found:
```
... this folder share the same name&#58; ${names}...
```
where the character sequence doesn't trip up the parser

or
```
... this folder share the same name${":"} ${names}...
```
which might be only slightly better because it's more obvious the problem with the `:` character.

But I could go either way, it's unfortunate that the parser doesn't interpret the backticks as quotes the same way.

What do you want to do?

jonpo replied:
html entity is the best option I think
